Modules/posixmodule.c

```diff
@@ -12287,6 +12296,37 @@ os_DirEntry_inode_impl(DirEntry *self)
 #endif
 }
 
+/*[clinic input]
+os.DirEntry.__reduce__
+
+returns null and raises an exception to avoid pickling
+[clinic start generated code]*/
+
+static PyObject *
+os_DirEntry___reduce___impl(DirEntry *self)
```

Comment:
Why this is needed? Does not `object.__reduce__` raise an error by default for non-trivial objects (if `obj->ob_type->tp_basicsize` > the basicsize including `__dict__`, `__weakref__` and slots)? See Objects/typeobject.c:4183.

Reply:
Ah! It's interesting that you point that out, I actually went through your [bpo](https://bugs.python.org/issue22995) before submitting this change!

Yes, you are completely right, this is not needed and it would still work. However, it is easy to construct a scenario where that won't work. For example, if `DirEntry` subclassed from `PyListType` or from a base with a custom `__reduce__`.

Given that I provided a custom `tp_new` slot, it is not immediately clear now that the type won't be pickable. I think that instead of relying on the internals of the object's `tp_base.__reduce__`, it is best if the user is explicit about the intent. In this specific case, I wanted to express that both `DirEntry` and `ScandirIterator` are not pickable.

 Imagine someone in the future migrating from PyType_Ready to PyType_FromSpec, only to realize that their instance is now pickable. Using this as an example, it will make sure that future users are more explicit about the intent.